File: bindings/php/ExecutionMode.php
<?php
namespace Stencil;

final class ExecutionMode {
    /** @internal */
    public \FFI\CData $ptr;
    private bool $owned;

    /** @internal */
    public function __construct(\FFI\CData $ptr, bool $owned) {
        $this->ptr = $ptr;
        $this->owned = $owned;
    }

    public static function fixedTicks( $ticks) {
        $ret = Lib::ffi()->ExecutionMode_fixed_ticks($ticks);
        return new ExecutionMode($ret, true);
    }

    public static function untilCondition(string $output_name,  $condition,  $max_ticks,  $check_interval) {
        $__n0 = strlen($output_name);
        $__view0 = Lib::ffi()->new('DiplomatStringView');
        if ($__n0 > 0) {
            $__buf0 = Lib::ffi()->new("uint8_t[" . $__n0 . "]", false);
            \FFI::memcpy($__buf0, $output_name, $__n0);
            $__view0->data = Lib::ffi()->cast('const char*', \FFI::addr($__buf0[0]));
        } else {
            $__view0->data = null;
        }
        $__view0->len = $__n0;
        $result = Lib::ffi()->ExecutionMode_until_condition($__view0, $condition->ptr, $max_ticks, $check_interval);
        if (!$result->is_ok) {
            throw new DiplomatError('NucleationError', $result->err, NucleationError::name($result->err));
        }
        return new ExecutionMode($result->ok, true);
    }

    public static function untilChange( $max_ticks,  $check_interval) {
        $ret = Lib::ffi()->ExecutionMode_until_change($max_ticks, $check_interval);
        return new ExecutionMode($ret, true);
    }

    public static function untilStable( $stable_ticks,  $max_ticks) {
        $ret = Lib::ffi()->ExecutionMode_until_stable($stable_ticks, $max_ticks);
        return new ExecutionMode($ret, true);
    }

    public function __destruct() {
        if ($this->owned) {
            Lib::ffi()->ExecutionMode_destroy($this->ptr);
        }
    }
}

File: bindings/kotlin/src/main/kotlin/at/schem/nucleation/ExecutionMode.kt
package at.schem.nucleation;
import com.sun.jna.Callback
import com.sun.jna.Library
import com.sun.jna.Native
import com.sun.jna.Pointer
import com.sun.jna.Structure

internal interface ExecutionModeLib: Library {
    fun ExecutionMode_destroy(handle: Pointer)
    fun ExecutionMode_fixed_ticks(ticks: FFIUint32): Pointer
    fun ExecutionMode_until_condition(outputName: Slice, condition: Pointer, maxTicks: FFIUint32, checkInterval: FFIUint32): ResultPointerInt
    fun ExecutionMode_until_change(maxTicks: FFIUint32, checkInterval: FFIUint32): Pointer
    fun ExecutionMode_until_stable(stableTicks: FFIUint32, maxTicks: FFIUint32): Pointer
}
/** How long to run the circuit for one `execute` call (PORTING rule 10).
*/
class ExecutionMode internal constructor (
    internal val handle: Pointer,
    // These ensure that anything that is borrowed is kept alive and not cleaned
    // up by the garbage collector.
    internal val selfEdges: List<Any>,
    internal var owned: Boolean,
)  {

    init {
        if (this.owned) {
            this.registerCleaner()
        }
    }

    private class ExecutionModeCleaner(val handle: Pointer, val lib: ExecutionModeLib) : Runnable {
        override fun run() {
            lib.ExecutionMode_destroy(handle)
        }
    }
    private fun registerCleaner() {
        CLEANER.register(this, ExecutionMode.ExecutionModeCleaner(handle, ExecutionMode.lib));
    }

    companion object {
        internal val libClass: Class<ExecutionModeLib> = ExecutionModeLib::class.java
        internal val lib: ExecutionModeLib = Native.load("nucleation", libClass)
        @JvmStatic
        
        /** Run for exactly `ticks` ticks.
        */
        fun fixedTicks(ticks: UInt): ExecutionMode {
            
            val returnVal = lib.ExecutionMode_fixed_ticks(FFIUint32(ticks));
            val selfEdges: List<Any> = listOf()
            val handle = returnVal 
            val returnOpaque = ExecutionMode(handle, selfEdges, true)
            return returnOpaque
        }
        @JvmStatic
        
        /** Run until the output named `output_name` meets `condition`,
        *checking every `check_interval` ticks, giving up after `max_ticks`
        *ticks (the result's `condition_met` reports which happened).
        */
        fun untilCondition(outputName: String, condition: OutputCondition, maxTicks: UInt, checkInterval: UInt): Result<ExecutionMode> {
            val outputNameSliceMemory = PrimitiveArrayTools.borrowUtf8(outputName)
            
            val returnVal = lib.ExecutionMode_until_condition(outputNameSliceMemory.slice, condition.handle, FFIUint32(maxTicks), FFIUint32(checkInterval));
            try {
                val nativeOkVal = returnVal.getNativeOk();
                if (nativeOkVal != null) {
                    val selfEdges: List<Any> = listOf()
                    val handle = nativeOkVal 
                    val returnOpaque = ExecutionMode(handle, selfEdges, true)
                    return returnOpaque.ok()
                } else {
                    return NucleationErrorError(NucleationError.fromNative(returnVal.getNativeErr()!!)).err()
                }
            } finally {
                outputNameSliceMemory.close()
            }
        }
        @JvmStatic
        
        /** Run until any output changes from its initial reading, checking
        *every `check_interval` ticks, giving up after `max_ticks` ticks.
        */
        fun untilChange(maxTicks: UInt, checkInterval: UInt): ExecutionMode {
            
            val returnVal = lib.ExecutionMode_until_change(FFIUint32(maxTicks), FFIUint32(checkInterval));
            val selfEdges: List<Any> = listOf()
            val handle = returnVal 
            val returnOpaque = ExecutionMode(handle, selfEdges, true)
            return returnOpaque
        }
        @JvmStatic
        
        /** Run (one tick at a time) until all outputs have been unchanged for
        *`stable_ticks` consecutive ticks, giving up after `max_ticks`
        *ticks (the result's `condition_met` reports stability).
        */
        fun untilStable(stableTicks: UInt, maxTicks: UInt): ExecutionMode {
            
            val returnVal = lib.ExecutionMode_until_stable(FFIUint32(stableTicks), FFIUint32(maxTicks));
            val selfEdges: List<Any> = listOf()
            val handle = returnVal 
            val returnOpaque = ExecutionMode(handle, selfEdges, true)
            return returnOpaque
        }
    }

}